[Bug]: panic: runtime error: invalid memory address or nil pointer dereference

### Terraform Core Version

1.9.2.

### ONTAP Provider Version

v1.1.1

### Affected Resource(s)

try to provisioning a igroup with initiators as below:

# Create a protocols_san_igroup
resource "netapp-ontap_protocols_san_igroup_resource" "protocols_san_igroups" {
  cx_profile_name = "ontap_lab"
  name = var.igroup_name
  svm = {
    name = var.svm_name
   }
  initiators = [{
     name = "iqn.2005-03.org.open-iscsi:host-246e965dc31c"
   }]
  os_type = "linux"
  comment = "test"
}


### Expected Behavior

igroup created

### Actual Behavior

Stack trace from the terraform-provider-netapp-ontap_v1.1.1 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xb2a98a]

goroutine 31 [running]:
github.com/netapp/terraform-provider-netapp-ontap/internal/provider.(*ProtocolsSanIgroupResource).Create(0xc0003c6880, {0xebaa58, 0xc00054a7e0}, {{{{0xebf518, 0xc000566c90}, {474c0, 0xc00054be00}}, {0xec3af8, 0xc000155130}}, {{{0xebf518, ...}, ...}, ...}, ...}, ...)
        github.com/netapp/terraform-provider-netapp-ontap/internal/provider/protocols_san_igroup_resource.go:258 +0x20a
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).CreateResource(0xc000320000, {0xebaa58, 0xc00054a7e0}, 0xc0004ed580, 0xc0004ed520)
        github.com/hashicorp/terraform-plugin-framework@v1.7.0/internal/fwserver/server_createresource.go:101 +0x578
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).ApplyResourceChange(0xc0005656d8?, {0xebaa58, 0xc00054a7e0}, 0xc00052ae10, 0xc0004ed6d8)
        github.com/hashicorp/terraform-plugin-framework@v1.7.0/internal/fwserver/server_applyresourcechange.go:57 +0x4a5
github.com/hashicorp/terraform-plugin-framework/internal/proto6server.(*Server).ApplyResourceChange(0xc000320000, {0xebaa58?, 0xc00054a6c0?}, 0xc00052ad70)
        github.com/hashicorp/terraform-plugin-framework@v1.7.0/internal/proto6server/server_applyresourcechange.go:55 +0x3e5
github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server).ApplyResourceChange(0xc0002205a0, {0xebaa58?, 0xc000547ce0?}, 0xc0001e24d0)
        github.com/hashicorp/terraform-plugin-go@v0.22.1/tfprotov6/tf6server/server.go:846 +0x3d0
github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_ApplyResourceChange_Handler({0xd2d500?, 0xc0002205a0}, {0xebaa58, 0xc000547ce0}, 0xc00018aa00, )
        github.com/hashicorp/terraform-plugin-go@v0.22.1/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:518 +0x169
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00019f000, {0xebaa58, 0xc000547c50}, {0xec2100, 0xc000007ba0}, 0xc000543440, 0xc000310f90, 0x13fdef8, 0x0)
        google.golang.org/grpc@v1.62.1/server.go:1386 +0xe23
google.golang.org/grpc.(*Server).handleStream(0xc00019f000, {0xec2100, 0xc000007ba0}, 0xc000543440)
        google.golang.org/grpc@v1.62.1/server.go:1797 +0x100c
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.62.1/server.go:1027 +0x8b
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 9
        google.golang.org/grpc@v1.62.1/server.go:1038 +0x135

Error: The terraform-provider-netapp-ontap_v1.1.1 plugin crashed!



### Relevant Error/Panic Output Snippet

```shell
Error: The terraform-provider-netapp-ontap_v1.1.1 plugin crashed!
```


### Terraform Configuration Files

Error: The terraform-provider-netapp-ontap_v1.1.1 plugin crashed!



### Steps to Reproduce

# Create a protocols_san_igroup
resource "netapp-ontap_protocols_san_igroup_resource" "protocols_san_igroups" {
  cx_profile_name = "ontap_lab"
  name = var.igroup_name
  svm = {
    name = var.svm_name
   }
  initiators = [{
     name = "iqn.2005-03.org.open-iscsi:host-246e965dc31c"
   }]
  os_type = "linux"
  comment = "test"
}


### Debug Output

_No response_

### Panic Output

_No response_

### Important Factoids

_No response_

### References

_No response_

### Would you like to implement a fix?

None

--- a/.github/workflows/acc_test.yml
+++ b/.github/workflows/acc_test.yml
@@ -25,6 +25,7 @@ jobs:
           TF_ACC_NETAPP_HOST2: ${{ secrets.TF_ACC_NETAPP_HOST2 }}
           TF_ACC_NETAPP_HOST3: ${{ secrets.TF_ACC_NETAPP_HOST3 }}
           TF_ACC_NETAPP_HOST4: ${{ secrets.TF_ACC_NETAPP_HOST4 }}
+          TF_ACC_NETAPP_HOST5: ${{ secrets.TF_ACC_NETAPP_HOST5 }}
           TF_ACC_NETAPP_HOST_CIFS: ${{ secrets.TF_ACC_NETAPP_HOST_CIFS }}
           TF_ACC_NETAPP_USER: ${{ secrets.TF_ACC_NETAPP_USER }}
           TF_ACC_NETAPP_PASS: ${{ secrets.TF_ACC_NETAPP_PASS }}

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,9 @@
+## 1.1.3
+
+BUG FIXES:
+* **netapp-ontap_protocols_san_igroups_resource**: fixed bug nil pointer dereference ([#247](https://github.com/NetApp/terraform-provider-netapp-ontap/issues/247))
+
+
 ## 1.1.2 (2024-06-03)
 
 ENHANCEMENTS:

--- a/internal/interfaces/protocols_san_igroup.go
+++ b/internal/interfaces/protocols_san_igroup.go
@@ -56,14 +56,14 @@ type Portset struct {
 
 // ProtocolsSanIgroupResourceBodyDataModelONTAP describes the body data model using go types for mapping.
 type ProtocolsSanIgroupResourceBodyDataModelONTAP struct {
-	Name       string            `mapstructure:"name"`
-	SVM        SvmDataModelONTAP `mapstructure:"svm"`
-	OsType     string            `mapstructure:"os_type"`
-	Protocol   string            `mapstructure:"protocol"`
-	Comment    string            `mapstructure:"comment,omitempty"`
-	Igroups    []IgroupLun       `mapstructure:"igroups,omitempty"`
-	Initiators []IgroupInitiator `mapstructure:"initiators,omitempty"`
-	Portset    Portset           `mapstructure:"portset,omitempty"`
+	Name       string                   `mapstructure:"name"`
+	SVM        SvmDataModelONTAP        `mapstructure:"svm"`
+	OsType     string                   `mapstructure:"os_type"`
+	Protocol   string                   `mapstructure:"protocol"`
+	Comment    string                   `mapstructure:"comment,omitempty"`
+	Igroups    []map[string]interface{} `mapstructure:"igroups,omitempty"`
+	Initiators []map[string]interface{} `mapstructure:"initiators,omitempty"`
+	Portset    Portset                  `mapstructure:"portset,omitempty"`
 }
 
 // UpdateProtocolsSanIgroupResourceBodyDataModelONTAP describes the body data model using go types for mapping.

--- a/internal/provider/protocols_san_igroup_resource.go
+++ b/internal/provider/protocols_san_igroup_resource.go
@@ -16,6 +16,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-framework/types"
 	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
 	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/mitchellh/mapstructure"
 	"github.com/netapp/terraform-provider-netapp-ontap/internal/interfaces"
 	"github.com/netapp/terraform-provider-netapp-ontap/internal/utils"
 )
@@ -40,16 +41,16 @@ type ProtocolsSanIgroupResource struct {
 
 // ProtocolsSanIgroupResourceModel describes the resource data model.
 type ProtocolsSanIgroupResourceModel struct {
-	CxProfileName types.String                                `tfsdk:"cx_profile_name"`
-	Name          types.String                                `tfsdk:"name"`
-	SVM           SVM                                         `tfsdk:"svm"`
-	Comment       types.String                                `tfsdk:"comment"`
-	Igroups       *[]ProtocolsSanIgroupResourceIgroupModel    `tfsdk:"igroups"`
-	Initiators    *[]ProtocolsSanIgroupResourceInitiatorModel `tfsdk:"initiators"`
-	OsType        types.String                                `tfsdk:"os_type"`
-	Portset       types.Object                                `tfsdk:"portset"`
-	Protocol      types.String                                `tfsdk:"protocol"`
-	ID            types.String                                `tfsdk:"id"`
+	CxProfileName types.String                               `tfsdk:"cx_profile_name"`
+	Name          types.String                               `tfsdk:"name"`
+	SVM           SVM                                        `tfsdk:"svm"`
+	Comment       types.String                               `tfsdk:"comment"`
+	Igroups       []ProtocolsSanIgroupResourceIgroupModel    `tfsdk:"igroups"`
+	Initiators    []ProtocolsSanIgroupResourceInitiatorModel `tfsdk:"initiators"`
+	OsType        types.String                               `tfsdk:"os_type"`
+	Portset       types.Object                               `tfsdk:"portset"`
+	Protocol      types.String                               `tfsdk:"protocol"`
+	ID            types.String                               `tfsdk:"id"`
 }
 
 // ProtocolsSanIgroupResourceIgroupModel describes the data source data model.
@@ -254,23 +255,35 @@ func (r *ProtocolsSanIgroupResource) Create(ctx context.Context, req resource.Cr
 	if !data.Comment.IsUnknown() {
 		body.Comment = data.Comment.ValueString()
 	}
-	if data.Initiators != nil {
-		body.Igroups = make([]interfaces.IgroupLun, len(*data.Igroups))
-		for index, record := range *data.Igroups {
-			body.Igroups[index] = interfaces.IgroupLun{
-				Name: record.Name.ValueString(),
-			}
+
+	if data.Igroups != nil {
+		igroups := []interfaces.IgroupLun{}
+		for _, igroup := range data.Igroups {
+			var ig interfaces.IgroupLun
+			ig.Name = igroup.Name.ValueString()
+			igroups = append(igroups, ig)
+		}
+		err := mapstructure.Decode(igroups, &body.Igroups)
+		if err != nil {
+			errorHandler.MakeAndReportError("error creating igroups", fmt.Sprintf("error on encoding copies info: %s, copies %#v", err, igroups))
+			return
 		}
 	}
 
 	if data.Initiators != nil {
-		body.Initiators = make([]interfaces.IgroupInitiator, len(*data.Initiators))
-		for index, record := range *data.Initiators {
-			body.Initiators[index] = interfaces.IgroupInitiator{
-				Name: record.Name.ValueString(),
-			}
+		initiators := []interfaces.IgroupInitiator{}
+		for _, v := range data.Igroups {
+			var initiator interfaces.IgroupInitiator
+			initiator.Name = v.Name.ValueString()
+			initiators = append(initiators, initiator)
+		}
+		err := mapstructure.Decode(initiators, &body.Initiators)
+		if err != nil {
+			errorHandler.MakeAndReportError("error creating igroups", fmt.Sprintf("error on encoding copies info: %s, copies %#v", err, initiators))
+			return
 		}
 	}
+
 	body.OsType = data.OsType.ValueString()
 	if !data.Portset.IsUnknown() {
 		var portset ProtocolsSanIgroupResourcePortsetModel
@@ -343,7 +356,7 @@ func (r *ProtocolsSanIgroupResource) Update(ctx context.Context, req resource.Up
 		request.OsType = data.OsType.ValueString()
 	}
 
-	tflog.Debug(ctx, fmt.Sprintf("update a svm resource: %#v", data))
+	tflog.Debug(ctx, fmt.Sprintf("update an igroup resource: %#v", data))
 	err = interfaces.UpdateProtocolsSanIgroup(errorHandler, *client, request, state.ID.ValueString())
 	if err != nil {
 		return

--- a/internal/provider/snapmirror_resource_test.go
+++ b/internal/provider/snapmirror_resource_test.go
@@ -23,7 +23,7 @@ func TestAccSnapmirrorResource(t *testing.T) {
 			{
 				Config: testAccSnapmirrorResourceBasicConfig("tf_peer:snap_source2", "terraform:snap_dest2"),
 				Check: resource.ComposeTestCheckFunc(
-					resource.TestCheckResourceAttr("netapp-ontap_snapmirror.example", "destination_endpoint.path", "terraform:snap_dest2"),
+					resource.TestCheckResourceAttr("netapp-ontap_snapmirror_resource.example", "destination_endpoint.path", "terraform:snap_dest2"),
 				),
 			},
 			// Update a policy
@@ -31,7 +31,7 @@ func TestAccSnapmirrorResource(t *testing.T) {
 				Config: testAccSnapmirrorResourceUpdateConfig("tf_peer:snap_source", "terraform:snap_dest", "MirrorAndVault"),
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("netapp-ontap_snapmirror_resource.example", "policy.name", "MirrorAndVault"),
-					resource.TestCheckResourceAttr("netapp-ontap_snapmirror.example", "destination_endpoint.path", "terraform:snap_dest"),
+					resource.TestCheckResourceAttr("netapp-ontap_snapmirror_resource.example", "destination_endpoint.path", "terraform:snap_dest"),
 				),
 			},
 			// Import and read
@@ -40,7 +40,7 @@ func TestAccSnapmirrorResource(t *testing.T) {
 				ImportState:   true,
 				ImportStateId: fmt.Sprintf("%s,%s", "terraform:snap_dest", "cluster4"),
 				Check: resource.ComposeTestCheckFunc(
-					resource.TestCheckResourceAttr("netapp-ontap_snapmirror.example", "destination_endpoint.path", "terraform:snap_dest"),
+					resource.TestCheckResourceAttr("netapp-ontap_snapmirror_resource.example", "destination_endpoint.path", "terraform:snap_dest"),
 				),
 			},
 			// Delete testing automatically occurs in TestCase

--- a/internal/provider/snapmirror_resource_test.go
+++ b/internal/provider/snapmirror_resource_test.go
@@ -16,31 +16,31 @@ func TestAccSnapmirrorResource(t *testing.T) {
 		Steps: []resource.TestStep{
 			// Test non existant Vol
 			{
-				Config:      testAccSnapmirrorResourceBasicConfig("snapmirror_dest_svm:testme", "snapmirror_dest_svm:testme"),
+				Config:      testAccSnapmirrorResourceBasicConfig("tf_peer:testme", "terraform:testme"),
 				ExpectError: regexp.MustCompile("6619337"),
 			},
 			// Create snapmirror and read
 			{
-				Config: testAccSnapmirrorResourceBasicConfig("snapmirror_dest_svm:snap_dest3", "snapmirror_dest_svm:snap_dest2"),
+				Config: testAccSnapmirrorResourceBasicConfig("tf_peer:snap_source2", "terraform:snap_dest2"),
 				Check: resource.ComposeTestCheckFunc(
-					resource.TestCheckResourceAttr("netapp-ontap_snapmirror_resource.example", "destination_endpoint.path", "snapmirror_dest_svm:snap_dest2"),
+					resource.TestCheckResourceAttr("netapp-ontap_snapmirror.example", "destination_endpoint.path", "terraform:snap_dest2"),
 				),
 			},
 			// Update a policy
 			{
-				Config: testAccSnapmirrorResourceUpdateConfig("snapmirror_dest_svm:snap_dest", "snapmirror_source_svm:snap", "MirrorAndVault"),
+				Config: testAccSnapmirrorResourceUpdateConfig("tf_peer:snap_source", "terraform:snap_dest", "MirrorAndVault"),
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("netapp-ontap_snapmirror_resource.example", "policy.name", "MirrorAndVault"),
-					resource.TestCheckResourceAttr("netapp-ontap_snapmirror_resource.example", "destination_endpoint.path", "snapmirror_source_svm:snap"),
+					resource.TestCheckResourceAttr("netapp-ontap_snapmirror.example", "destination_endpoint.path", "terraform:snap_dest"),
 				),
 			},
 			// Import and read
 			{
 				ResourceName:  "netapp-ontap_snapmirror_resource.example",
 				ImportState:   true,
-				ImportStateId: fmt.Sprintf("%s,%s", "snapmirror_dest_svm:snap_dest2", "cluster4"),
+				ImportStateId: fmt.Sprintf("%s,%s", "terraform:snap_dest", "cluster4"),
 				Check: resource.ComposeTestCheckFunc(
-					resource.TestCheckResourceAttr("netapp-ontap_snapmirror_resource.example", "destination_endpoint.path", "snapmirror_dest_svm:snap_dest2"),
+					resource.TestCheckResourceAttr("netapp-ontap_snapmirror.example", "destination_endpoint.path", "terraform:snap_dest"),
 				),
 			},
 			// Delete testing automatically occurs in TestCase
@@ -49,11 +49,11 @@ func TestAccSnapmirrorResource(t *testing.T) {
 }
 
 func testAccSnapmirrorResourceBasicConfig(sourceEndpoint string, destinationEndpoint string) string {
-	host := os.Getenv("TF_ACC_NETAPP_HOST4")
+	host := os.Getenv("TF_ACC_NETAPP_HOST5")
 	admin := os.Getenv("TF_ACC_NETAPP_USER")
-	password := os.Getenv("TF_ACC_NETAPP_PASS")
+	password := os.Getenv("TF_ACC_NETAPP_PASS2")
 	if host == "" || admin == "" || password == "" {
-		fmt.Println("TF_ACC_NETAPP_HOST4, TF_ACC_NETAPP_USER, and TF_ACC_NETAPP_PASS must be set for acceptance tests")
+		fmt.Println("TF_ACC_NETAPP_HOST5, TF_ACC_NETAPP_USER, and TF_ACC_NETAPP_PASS must be set for acceptance tests")
 		os.Exit(1)
 	}
 	return fmt.Sprintf(`
@@ -81,11 +81,11 @@ resource "netapp-ontap_snapmirror_resource" "example" {
 }
 
 func testAccSnapmirrorResourceUpdateConfig(sourceEndpoint string, destinationEndpoint string, policy string) string {
-	host := os.Getenv("TF_ACC_NETAPP_HOST4")
+	host := os.Getenv("TF_ACC_NETAPP_HOST5")
 	admin := os.Getenv("TF_ACC_NETAPP_USER")
-	password := os.Getenv("TF_ACC_NETAPP_PASS")
+	password := os.Getenv("TF_ACC_NETAPP_PASS2")
 	if host == "" || admin == "" || password == "" {
-		fmt.Println("TF_ACC_NETAPP_HOST4, TF_ACC_NETAPP_USER, and TF_ACC_NETAPP_PASS must be set for acceptance tests")
+		fmt.Println("TF_ACC_NETAPP_HOST5, TF_ACC_NETAPP_USER, and TF_ACC_NETAPP_PASS must be set for acceptance tests")
 		os.Exit(1)
 	}
 	return fmt.Sprintf(`

--- a/internal/provider/svm_peers_resource_test.go
+++ b/internal/provider/svm_peers_resource_test.go
@@ -39,7 +39,7 @@ func TestAccSvmPeersResource(t *testing.T) {
 			{
 				ResourceName:  "netapp-ontap_svm_peers_resource.example",
 				ImportState:   true,
-				ImportStateId: fmt.Sprintf("%s,%s,%s,%s", "snapmirror_dest_dp", "snapmirror_dest_svm", "swenjuncluster-1", "cluster4"),
+				ImportStateId: fmt.Sprintf("%s,%s,%s,%s", "terraform", "tf_peer", "swenjuncluster-1", "cluster4"),
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("netapp-ontap_svm_peers_resource.example", "svm.name", "snapmirror_dest_dp"),
 				),
@@ -48,13 +48,13 @@ func TestAccSvmPeersResource(t *testing.T) {
 	})
 }
 func testAccSvmPeersResourceConfig(svm, peerSvm, peerCluster, applications string) string {
-	host := os.Getenv("TF_ACC_NETAPP_HOST4")
+	host := os.Getenv("TF_ACC_NETAPP_HOST5")
 	admin := os.Getenv("TF_ACC_NETAPP_USER")
 	password := os.Getenv("TF_ACC_NETAPP_PASS")
 	password2 := os.Getenv("TF_ACC_NETAPP_PASS2")
 	host2 := os.Getenv("TF_ACC_NETAPP_HOST2")
 	if host == "" || admin == "" || password == "" {
-		fmt.Println("TF_ACC_NETAPP_HOST2, TF_ACC_NETAPP_HOST4, TF_ACC_NETAPP_USER, TF_ACC_NETAPP_PASS2 and TF_ACC_NETAPP_PASS must be set for acceptance tests")
+		fmt.Println("TF_ACC_NETAPP_HOST2, TF_ACC_NETAPP_HOST5, TF_ACC_NETAPP_USER, TF_ACC_NETAPP_PASS2 and TF_ACC_NETAPP_PASS must be set for acceptance tests")
 		os.Exit(1)
 	}
 	return fmt.Sprintf(`
@@ -92,5 +92,5 @@ resource "netapp-ontap_svm_peers_resource" "example" {
     peer_cx_profile_name = "cluster3"
   }
   applications = ["%s"]
-}`, host, admin, password, host2, admin, password2, svm, peerSvm, peerCluster, applications)
+}`, host, admin, password2, host2, admin, password2, svm, peerSvm, peerCluster, applications)
 }

